google/cloud/pubsub_v1/subscriber/_protocol/dispatcher.py

```diff
@@ -179,16 +182,35 @@ def ack(self, items: Sequence[requests.AckRequest]) -> None:
         # We must potentially split the request into multiple smaller requests
         # to avoid the server-side max request size limit.
         items_gen = iter(items)
-        ack_ids_gen = (item.ack_id for item in items)
         total_chunks = int(math.ceil(len(items) / _ACK_IDS_BATCH_SIZE))
 
+        exactly_once_delivery_enabled = self._manager._exactly_once_delivery_enabled()
         for _ in range(total_chunks):
-            ack_reqs_dict = {
-                req.ack_id: req
-                for req in itertools.islice(items_gen, _ACK_IDS_BATCH_SIZE)
-            }
+            ack_reqs_dict = {}
+            ack_ids = []
+            for req in itertools.islice(items_gen, _ACK_IDS_BATCH_SIZE):
+                # If this is a duplicate AckRequest, we do not pass it to send_unary_ack
+                if req.ack_id in ack_reqs_dict:
```

Comment:
We need to de-duplicate for modack and nack as well. For both, return an exception for their futures if there's duplicates.

I wonder if our own leaser might generate multiple modacks in one batch. The dispatcher request batching timer is [0.01 seconds](https://github.com/googleapis/python-pubsub/blob/0a11e078660eb7213a1a323fdf701d802fb1149c/google/cloud/pubsub_v1/subscriber/_protocol/dispatcher.py#L55). But the leaser might wait [0 seconds](https://github.com/googleapis/python-pubsub/blob/0a11e078660eb7213a1a323fdf701d802fb1149c/google/cloud/pubsub_v1/subscriber/_protocol/leaser.py#L206) before re-leasing messages. The solution is to increase leaser min wait time between runs to at least, 0.01 seconds (aka the dispatcher's _MAX_BATCH_LATENCY).

Now that I think of it, it's bonkers that the [time between lease runs](https://github.com/googleapis/python-pubsub/blob/0a11e078660eb7213a1a323fdf701d802fb1149c/google/cloud/pubsub_v1/subscriber/_protocol/leaser.py#L206) is `random.uniform(0.0, deadline * 0.9)`. Filed a bug for this: https://github.com/googleapis/python-pubsub/issues/675